### ISSUE TITLE
Create initial raw databases

### ISSUE THREAD

**Issue by yoid2000** [gda-score/anon-methods#1]
For your first task, I'd like you to create some databases for me, to be installed on a postgres server. The server credentials will be sent to you by Mickler (this was the prior email I just sent).

For background, please read https://aircloak.com/wp-content/uploads/Aircloak-Challenge.pdf

In there you will see a description of the databases that were used for our attack challenge. What I want to do is to make new versions of four of those databases, but with the following differences:

1. They will (most of them) be smaller,
2. They will have some additional columns consisting of personal information.
3. A few column and table names will be changed

The four databases you should work from are taxi, census1, banking, and scihub. These are already on the server for which you will get the credentials.

From each of the four, you will make three database with the following naming scheme:

-  raw_db
-  raw_db_link
-  raw_db_link_pub

Where 'db' is one of 'taxi', 'census', 'banking', and 'scihub'.

Hopefully it is clear from the challenge rules how each is constructed.

Regarding shrinking:

-  For the taxi database, use four hours only, say 8AM to noon
-  For the census database, use 250000 rows only
-  For scihub, use one week from sep2015

Regarding extra columns:

You can see that the current databases all have a column 'lastname'.

In addition to the 'lastname' column, I want to add the following columns:

firstname
birthdate
gender
ssn (social security number)
email
street
zip

In other words, everywhere the 'lastname' column currently exists, add these columns.

You can generate these columns from the generatedata tool (I'll have the credentials sent to you).

Regarding change of column name, on the banking database, please change 'client_id' to 'uid' everywhere it appears. On the taxi database, please change 'hack' to 'uid'.

Regarding change of table name, for the taxi database, please name the table 'rides'. For the census database, please name the table 'persons'. For the scihub database, please call the table 'downloads'. 

I doubt that these instructions are 100% clear, so please let me know if you have questions.

**Comment by ku294714** [comment]
I have gone through the databases and tried to understand the working. Does the query system of the web interface:
https://demo.aircloak.com/data_sources/taxi
does not work?

I tried querying the records of a few of the tables but it does not show up. It goes till processing state and then gets stuck at that point.

Can you please help me with this?

Regards,
Kumar Ankit

**Comment by yoid2000** [comment]
What query did you run?

Note also that for this particular issue, you don't need to query Aircloak. So after all this question could have been sent by email  ;)

**Comment by yoid2000** [comment]
Also what do you mean by "gets stuck"? I just ran `select count(*) from jan08` and it worked fine.

**Comment by ku294714** [comment]
I ran this query: 
select * from jan08 order by hack desc limit 10;
select count(*) would just return the number of records.

**Comment by yoid2000** [comment]
In any event, please try "select count(*) from jan08" just to confirm that
it works.

Your query is very inefficient, and so took a long time. It probably would
have finished eventually.

What your query did was transfer the whole database to the cloak, the ask
the cloak to order the whole database by hack. That's a lot of work for the
cloak so it would take a while.

A faster way to run the query would be this:

```
select * from (
  select * from jan08
  order by hack desc
  limit 10) t
```

That pushes all of the work into the database, where it can happen quickly.
Then the database returns only 10 rows to the cloak, which can then process
them fast.

**Comment by ku294714** [comment]
Got It. Thank you.

**Comment by yoid2000** [comment]
By the way, for building the tables for this issue, please use the files from here:

https://download.aircloak.com/

That's because there are problems with at least one of the databases on db001.gda-score.org

**Comment by ku294714** [comment]
Sir, 

I feel I do not have access to create Database on the DB server. Also, I do not have access to see the details of the tables created by you.

**Comment by yoid2000** [comment]
HI Ankit,

Please try again. I just assigned you write privileges (at least I think I
did).

And, what do you mean when you say you don't have access to see the details
of the tables I created?

PF

**Comment by ku294714** [comment]
While accessing the tables through aircloack, I can access them. However, if I want to access those relations from the DB server, I get the access denied message. For instance, for relation sep2015 under DB scihub, I get the error as permission denied for this relation.

Also, I can see that I can create the DBs now. 

Thank you.

**Comment by yoid2000** [comment]
What query do you use to get the relation?

**Comment by ku294714** [comment]
I used normal select query.

**Comment by yoid2000** [comment]
Please show me the exact query...

**Comment by ku294714** [comment]
select count(*) from public.sep2015;
select count(*) from sep2015;
select * from sep2015 order by uid limit 10;

Also, I tried exporting this table.

I tried the same under scihub, postgres, and scihub_link DBs.
----------------
Message: ERROR:  permission denied for relation sep2015
SQL state: 42501

**Comment by yoid2000** [comment]
Is anything working? I'm having no trouble with those queries at least
using pgadmin4 GUI.

PF

**Comment by ku294714** [comment]
I installed pgadmin4 GUI. I created my own table. I guess the table owner is the issue, maybe you can grant Select privilege to me?

**Comment by yoid2000** [comment]
I made a helpdesk ticket for our IT people. YOu were copied so you should
have seen it. Please communicate with Pascal on this.

**Comment by ku294714** [comment]
Hello Sir,

Yesterday, my access issue was resolved.
I have created the raw_scihub DB and created the table downloads and have added extra columns.
This table has the first week of data from sep2015.

I have used the data from the table sep2015 present under scihub DB.
Please check if everything is fine. I have granted you all the privileges of the table.

Also, I would like to draw attention to one point that the data in the table sep2015 under scihub DB differs from the raw files that you had provided, so I have used the table data.

Regards,
Ankit

**Comment by yoid2000** [comment]
On Fri, Sep 28, 2018 at 11:07 AM Ankit Dixit <notifications@github.com>
wrote:

> Hello Sir,
>
> Yesterday, my access issue was resolved.
> I have created the raw_scihub DB and created the table downloads and have
> added extra columns.
>
The extra columns are currently NULL. I assume you plan to fill them in???

> This table has the first week of data from sep2015.
>
> I have used the data from the table sep2015 present under scihub DB.
> Please check if everything is fine. I have granted you all the privileges
> of the table.
>
> Also, I would like to draw attention to one point that the data in the
> table sep2015 under scihub DB differs from the raw files that you had
> provided, so I have used the table data.
>

What was the difference?

PF

**Comment by ku294714** [comment]
What data shall I fill in? Is there any data available? Or I can fill in
some random data?

Regarding the difference, there is the count mismatch that I have seen. I
can check the difference if you want me to.

Regards,
Ankit

**Comment by yoid2000** [comment]
On Fri, Sep 28, 2018 at 3:02 PM Ankit Dixit <notifications@github.com>
wrote:

> What data shall I fill in? Is there any data available? Or I can fill in
> some random data?
>

The data you fill in should come from generatedata! Use generatedata for
the NULL columns, of course done so that any given UID has the same values
for each of the new columns in every row of the UID.

Let me know when you've done that for scihub, and if it looks ok after I
check it you can do the others.


>
> Regarding the difference, there is the count mismatch that I have seen. I
> can check the difference if you want me to.
>

No need to check why. How big was the count mismatch? (Thinking back, it
might be that I removed some garbage data, but I don't remember clearly
....)

PF

**Comment by ku294714** [comment]
Okay sir.

**Comment by ku294714** [comment]
Hello Sir,

I have created the table downloads under raw_scihub DB and also populated the data, please check if everything is fine and if I can proceed with the creation of remaining DBs.

**Comment by yoid2000** [comment]
You have a lot of tables there. Which one should I be looking at?

**Comment by ku294714** [comment]
Downloads

**Comment by yoid2000** [comment]
I had a look. I think it is fine. generatedata seems to create a small
fraction of duplicates, but that is not really a problem. Real data is
dirty anyway, so this is ok.

Please go ahead with the other databases!

**Comment by ku294714** [comment]
Okay sir.

**Comment by ku294714** [comment]
Hello Sir, 

I have also created raw_scihub_link and raw_scihub_link_pub DBs and both have 'downloads' tables with 1 week of data in it.

Please check. I am starting with the creation of taxi DBs now.

**Comment by yoid2000** [comment]
Hi Ankit,

I am leaving early tomorrow on a brief holiday, so probably won't have a
chance to check your work until I return Wednesday. Please continue with
the other DBs.

PF

**Comment by ku294714** [comment]
Hello Sir,

I have created the DBs with the instructions mentioned above.

However, for the Banking DBs, I added the extra columns in only 'Accounts' table. I feel as we already have uid present in other tables, we can fetch the details of these extra columns from accounts table itself and there is no need to add these redundant columns in all the tables. Please correct me if I am wrong.

Also, please check all the other DBs.

**Comment by yoid2000** [comment]
Please add the extra columns to all the tables in banking. I know it seems kind of silly, but I want to be able to treat each table as a stand-alone entity, and avoid the need for doing JOINs in the attacks themselves. I also understand that we could accomplish this with a view, but these are small tables and the extra storage needed to replicate the extra columns is not a factor. On the other hand, some attacks will take thousands of queries, and I'd rather avoid the cost of the JOIN on each one.

**Comment by yoid2000** [comment]
The "link" tables are constructed incorrectly. For a given uid, all of the values for the added columns should be the same between the three tables (the basic table, the link table, and the link_pub) table. 

For instance, here is a row from `raw_scihub`:

![image](https://user-images.githubusercontent.com/2162345/46940580-5ae68380-d069-11e8-831f-d2fd4f68af3b.png)

and here is a row for the same user from `raw_scihub_link`:

![image](https://user-images.githubusercontent.com/2162345/46940623-72be0780-d069-11e8-8cb3-7acb02d97bdf.png)

As you can see, the `firstname`, `birthdate`, `gender`, etc. are different. They should be the same. The whole point of the `link` and `link_pub` tables are that they represent a subset of the basic table, not a different table.

**Comment by ku294714** [comment]
Oh. Okay. I thought we shall generate the different data for all the tables.

I will update it.

Regards,
Ankit

**Comment by ku294714** [comment]
Sir,

I have updated the DBs, scihub, and taxi. Please check.

However, regarding the subset of the table, that means for census also, whatever data we have in the main table, we shall have in the link and link_pub tables.
But what I did was I selected the first 250000 records for the all the tables in the census, which resulted in the different set of records from the main table and subset tables.

And if I push similar records in the subset tables as they are in the census main table, there would be no difference in the subset tables.

I hope I am able to clearly put forward my point, or else I can come to discuss over this tomorrow.

Maybe this would help:

Row from raw_census:

select * from persons where uid='250101'   ---does not contain any data

However, the data for this exists under raw_census_link

**Comment by yoid2000** [comment]
Sorry if this question seems rude, but have you carefully read this document, and do you understand it?

https://aircloak.com/wp-content/uploads/Aircloak-Challenge.pdf

Because if you have you should understand the purpose of the link and link_pub tables, and therefore how they should be configured.

**Comment by ku294714** [comment]
I had gone through the document. I will go through again and try to correct the census DB.

**Comment by ku294714** [comment]
I have updated the banking, scihub, and taxi DBs. please check, meanwhile I am working on census DB

**Comment by yoid2000** [comment]
Updated how? It is still wrong. For instance, in the raw_banking.cards
table, uid 9 has firstname "Ira", while in the raw_banking_link.cards
table, uid 9 has firstname "Dante".

PF

**Comment by ku294714** [comment]
Hello Sir, 

I have tried to correct all the DBs and also checked them.

Can you also please verify.

Also, I tried to understand the linkability attack. 

link_pub tables are prior knowledge tables, and link tables are protected tables for linkability attacks.

"However, as we discussed last time, all the data which is present in the link and prior tables must be present in the original raw tables. I discussed the same with Ali, and he told there could be some data in these subset tables which might not be present in the main table." 

Is this understanding of mine with the linkability attack is correct?

Regards,
Ankit

**Comment by yoid2000** [comment]
I checked the tables and they look fine.

Could you clean up and then close the issue? By clean up, I mean remove the raw_census0 databases and remove all the extra tables (temp tables etc.).

**Comment by ku294714** [comment]
Okay sir. I will do the cleanup and close this issue.

**Comment by ku294714** [comment]
Done with the cleanup, closing the issue
